design/nodes-machines-and-hosts.md

```diff
@@ -0,0 +1,213 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# nodes-machines-and-hosts
+
+## Status
+
+implementable
+
+## Table of Contents
+
+<!--ts-->
+   * [nodes-machines-and-hosts](#nodes-machines-and-hosts)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+            * [Data Model](#data-model)
+            * [Growing the Cluster](#growing-the-cluster)
+            * [Shrinking the Cluster](#shrinking-the-cluster)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives](#alternatives)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-20T17:58-05:00 -->
+
+<!--te-->
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This document explains the use of Node, Machine, and BareMetalHost
+objects to represent different facets of implementing a self-hosted
+and Kubernetes-native hardware provisioning system. It explains the
+role of each, and why we are using 3 separate resources.
+
+## Motivation
+
+This document details the original design decisions made when the
+MetalKube project was launched, and places the information in context
+as an aid to future contributors.
+
+### Goals
+
+- Explain Node, Machine, and BareMetalHost and how they relate to
+  MetalKube.
+- Encourage input on that design, to refine it.
+
+### Non-Goals
+
+- Discussion of user interface designs related to the resources.
+
+## Proposal
+
+The MetalKube project is building a hardware provisioning system with
+a Kubernetes-native API. It includes a provider implementation that
+fits into the API defined by the cluster-api SIG. It differs from
+other similar implementations by being self-hosted. Where the AWS or
+OpenStack providers run within Kubernetes, they allocate resources
+from a pool *outside* of the Kubernetes cluster. MetalKube will use
```

Comment:
I may expand on "allocate resources from a pool outside of the Kubernetes cluster" to also clarify that this is done by talking to an API external to the cluster.  Maybe that's clear, but I want to hammer the point about the difference between the API being within Kubernetes or outside of it.

Reply:
I expanded that sentence to emphasize this point.